### DISCUSSION TITLE
No verbatim headers possible

### DISCUSSION THREAD

**Discussion by milabs** [earthboundkid/requests#51] [Ideas] [answered]
https://github.com/carlmjohnson/requests/blob/7485e76abb3cc927ed6c8a13389ebc3abb6a0ce9/builder_core.go#L232

It's not possible to have non-canonical headers with current Api. I'd suggest to add `VerbarimHeader()` or something like that.

**Comment by earthboundkid** [comment, marked as answer]
In theory, it shouldn't matter but in practice, some servers suck. I had this problem when I was using [Ruby to talk to some government server back in 2016](https://github.com/jnunemaker/httparty/issues/406), so I'm definitely sympathetic to the issue.

I hate to junk up the main interface for what's really a workaround for the server sucking. Maybe a custom transport? In most cases, I guess you probably just need all lowercase or something, so maybe just write


```go
LowercaseHeaders := requests.RoundTripFunc(func(req *http.Request) (res *http.Response, err error) {
	req2 := *req
	req2.Header = make(http.Header, len(req.Header))
	for k, v := range req.Header {
		req2.Header[strings.ToLower(k)] = v
	}
	return http.DefaultTransport.RoundTrip(&req2)
})
rb := requests.
	URL("https://example.com").
	Transport(LowercaseHeaders)
```

Will that work for you?

**Reply by milabs** [reply]
Hah, well, it's actually a nice wrapper and it's definitely would solve the issue. But as a user I personally prefer to have something for that in the library. Anyway, thanks for the tip!

UPDATE: 

```
rb := requests.
	VerbatimHeader("key", "val").
	URL("https://example.com")
```

UPDATE-2:

Or modify `Header()` so if the KEY starts with a space ... treat is as a verbatim?
